libraries/Adapters/Microsoft.Bot.Builder.Adapters.Twilio/TwilioHelper.cs

```diff
@@ -0,0 +1,142 @@
+﻿using System;
+using System.Collections.Generic;
+using System.IO;
+using Microsoft.AspNetCore.Http;
+using Microsoft.Bot.Schema;
+using Newtonsoft.Json;
+using Twilio.Exceptions;
+using Twilio.Rest.Api.V2010.Account;
+using Twilio.Security;
+
+namespace Microsoft.Bot.Builder.Adapters.Twilio
+{
+    public static class TwilioHelper
+    {
+        /// <summary>
+        /// Extracts attachments (if any) from a twilio message and returns them in an Attachments array.
+        /// </summary>
+        /// <param name="message">The TwilioEvent message.</param>
+        /// <returns>An Attachments array with the converted attachments.</returns>
+        public static List<Attachment> GetMessageAttachments(TwilioEvent message)
+        {
+            var attachments = new List<Attachment>();
+            if (int.TryParse(message.NumMedia, out var numMediaResult) && numMediaResult > 0)
+            {
+                for (var i = 0; i < numMediaResult; i++)
+                {
+                    var attachment = new Attachment()
+                    {
+                        ContentType = message.MediaContentTypes[i],
+                        ContentUrl = message.MediaUrls[i].AbsolutePath,
+                    };
+                    attachments.Add(attachment);
+                }
+            }
+
+            return attachments;
+        }
+
+        /// <summary>
+        /// Converts a query string to a dictionary with key-value pairs.
+        /// </summary>
+        /// <param name="query">The query string to convert.</param>
+        /// <returns>A dictionary with the query values.</returns>
+        public static Dictionary<string, string> QueryStringToDictionary(string query)
+        {
+            var pairs = query.Replace("+", "%20").Split('&');
+            var values = new Dictionary<string, string>();
+
+            foreach (var p in pairs)
+            {
+                var pair = p.Split('=');
+                var key = pair[0];
+                var value = Uri.UnescapeDataString(pair[1]);
+
+                values.Add(key, value);
+            }
+
+            return values;
+        }
+
+        /// <summary>
+        /// Formats a BotBuilder activity into an outgoing Twilio SMS message.
+        /// </summary>
+        /// <param name="activity">A BotBuilder Activity object.</param>
+        /// <param name="options">A set of params with the required values for authentication.</param>
+        /// <returns>A Message's options object with {body, from, to, mediaUrl}.</returns>
+        public static CreateMessageOptions ActivityToTwilio(Activity activity, ITwilioAdapterOptions options)
+        {
```

Comment:
We have to check if _activity_, _activity.ChannelData_ and _options_ are _**null**_. Do you think it's ok an empty _CreateMesssageOptions_ when any of the mentioned cases happen? Or should we return _null_?

Reply:
ChannelData is already being checked in that code, and the behavior is as expected. As for the options (involving the Twilio number) a warning might be preferable?

Reply:
I think the ChannelData could be null, It shouldn't return exception there, but if options is null the messageOptions shouldn't be created (it needs the TwilioNumber).